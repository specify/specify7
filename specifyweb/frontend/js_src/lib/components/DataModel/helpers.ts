import { f } from '../../utils/functools';
import type { RA, ValueOf } from '../../utils/types';
import { isTreeResource } from '../InitialContext/treeRanks';
import { relationshipIsToMany } from '../WbPlanView/mappingHelpers';
import type {
  AnySchema,
  AnyTree,
  SerializedResource,
  TableFields,
} from './helperTypes';
import type { SpecifyResource } from './legacyTypes';
import type { LiteralField, Relationship } from './specifyField';
import type { SpecifyTable } from './specifyTable';
import type { Tables } from './types';
/**
 * Lookups for operaters and relationships in Queries on the backend are
 * separated by `__`
 * The api supports the same syntax in query paramters.
 * For example `/api/specify/collectionobject/?collection__discipline__name__startswith="Invert"&catalognumber__gt=000000100`
 * fetches all collectionobjects in disciplines which names start with
 * "Invert" with catalognumbers greater than 100
 */
export const djangoLookupSeparator = '__';

export const backboneFieldSeparator = '.';

/**
 * Formats a relationship lookup which is used in a query passed to the backend.
 * For example `formatRelationshipPath('collection', 'discipline', 'division')`
 * becomes `'collection__discipline__division'`
 */
export const formatRelationshipPath = (...fields: RA<string>): string =>
  fields.join(djangoLookupSeparator);

const weekDayMap = {
  Sunday: 1,
  Monday: 2,
  Tuesday: 3,
  Wednesday: 4,
  Thursday: 5,
  Friday: 6,
  Saturday: 7,
};

/**
 * Use this to construct a query using a lookup for Django.
 * Returns an object which can be used as a filter when fetched from the backend.
 * Example: backendFilter('number1').isIn([1, 2, 3]) is the equivalent
 * of {number1__in: [1, 2, 3].join(',')}
 *
 * See the Django docs at:
 * https://docs.djangoproject.com/en/3.2/ref/models/querysets/#field-lookups
 */
export const backendFilter = (field: string) => ({
  equals: (value: number | string) => ({
    [[field, 'exact'].join(djangoLookupSeparator)]: value,
  }),
  contains: (value: string) => ({
    [[field, 'contains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveContains: (value: string) => ({
    [[field, 'icontains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveStartsWith: (value: string) => ({
    [[field, 'istartswith'].join(djangoLookupSeparator)]: value,
  }),
  startsWith: (value: string) => ({
    [[field, 'startswith'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveEndsWith: (value: string) => ({
    [[field, 'iendswith'].join(djangoLookupSeparator)]: value,
  }),
  endsWith: (value: string) => ({
    [[field, 'endswith'].join(djangoLookupSeparator)]: value,
  }),
  isIn: (value: RA<number | string>) => ({
    [[field, 'in'].join(djangoLookupSeparator)]: value.join(','),
  }),
  isNull: (value: 'false' | 'true' = 'true') => ({
    [[field, 'isnull'].join(djangoLookupSeparator)]: value,
  }),
  greaterThan: (value: number) => ({
    [[field, 'gt'].join(djangoLookupSeparator)]: value,
  }),
  greaterThanOrEqualTo: (value: number) => ({
    [[field, 'gte'].join(djangoLookupSeparator)]: value,
  }),
  lessThan: (value: number) => ({
    [[field, 'lt'].join(djangoLookupSeparator)]: value,
  }),
  lessThanOrEqualTo: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  matchesRegex: (value: string) => ({
    [[field, 'regex'].join(djangoLookupSeparator)]: value,
  }),

  dayEquals: (value: number) => ({
    [[field, 'day'].join(djangoLookupSeparator)]: value,
  }),
  monthEquals: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  yearEquals: (value: number) => ({
    [[field, 'year'].join(djangoLookupSeparator)]: value,
  }),
  weekEquals: (value: number) => ({
    [[field, 'week'].join(djangoLookupSeparator)]: value,
  }),
  weekDayEquals: (value: keyof typeof weekDayMap) => ({
    [[field, 'week_day'].join(djangoLookupSeparator)]: weekDayMap[value],
  }),
});

export const isResourceOfType = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
  // @ts-expect-error
): resource is SpecifyResource<Tables[TABLE_NAME]> =>
  resource.specifyTable.name === tableName;

export const toTable = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  resource.specifyTable.name === tableName
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

export const toResource = <TABLE_NAME extends keyof Tables>(
  resource: SerializedResource<AnySchema>,
  tableName: TABLE_NAME
): SerializedResource<Tables[TABLE_NAME]> | undefined =>
  resource._tableName === tableName
    ? (resource as SerializedResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * The table.field has a very broad type to reduce type conflicts in components
 * that deal with generic schemas (accept AnySchema or a SCHEMA extends AnySchema)
 */
export const getField = <
  SCHEMA extends ValueOf<Tables>,
  FIELD extends TableFields<SCHEMA>
>(
  table: SpecifyTable<SCHEMA>,
  name: FIELD
): FIELD extends keyof SCHEMA['fields'] ? LiteralField : Relationship =>
  table.field[name] as FIELD extends keyof SCHEMA['fields']
    ? LiteralField
    : Relationship;

export const toTreeTable = (
  resource: SpecifyResource<AnySchema>
): SpecifyResource<AnyTree> | undefined =>
  isTreeResource(resource) ? resource : undefined;

export const toTables = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableNames: RA<TABLE_NAME>
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  f.includes(tableNames, resource.specifyTable.name)
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * Example usage:
 * resource: Collector
 * fields: agent -> lastName
 * Would return [agent, lastName] if agent exists
 *
 */
export async function fetchDistantRelated(
  resource: SpecifyResource<AnySchema>,
  fields: RA<LiteralField | Relationship> | undefined
): Promise<
  | {
      readonly resource: SpecifyResource<AnySchema> | undefined;
      readonly field: LiteralField | Relationship | undefined;
    }
  | undefined
> {
  if (
    Array.isArray(fields) &&
    fields.some(
      (field) =>
        field.isRelationship &&
        relationshipIsToMany(field) &&
        field !== fields.at(-1)
    )
  ) {
    console.error(
      'Can not index inside of a -to-many relationship. Use an aggregator instead'
    );
    return undefined;
  }

  const related =
    fields === undefined || fields.length === 0
      ? resource
      : fields.length === 1
      ? await resource.fetch()
      : await resource.rgetPromise(
          fields
            .slice(0, -1)
            .map(({ name }) => name)
            .join(backboneFieldSeparator)
        );

  const field = fields?.at(-1);
  const relatedResource = related ?? undefined;
  return relatedResource === undefined && field === undefined
    ? undefined
    : {
        resource: relatedResource,
        field,
      };
}
