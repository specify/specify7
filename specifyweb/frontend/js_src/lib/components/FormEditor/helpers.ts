import type { RR } from '../../utils/types';
import type { Tables } from '../DataModel/types';

export const tablesWithBusRulesIn6 = new Set([
  'AccessionAgentBusRules',
  'AccessionAuthorizationBusRules',
  'AccessionBusRules',
  'AddressBusRules',
  'AgentBusRules',
  'AppraisalBusRules',
  'AttachmentBusRules',
  'AuthorBusRules',
  'BioStratBusRules',
  'BorrowBusRules',
  'CatalogNumberingSchemeBusRules',
  'CatAutoNumberingSchemeBusRules',
  'CollectingEventBusRules',
  'CollectingTripAuthorizationBusRules',
  'CollectingTripBusRules',
  'CollectionBusRules',
  'CollectionObjectBusRules',
  'CollectorBusRules',
  'ConservDescriptionBusRules',
  'ConservEventBusRules',
  'ContainerBusRules',
  'DataTypeBusRules',
  'DeaccessionAgentBusRules',
  'DeterminationBusRules',
  'DeterminerBusRules',
  'DisciplineBusRules',
  'DivisionBusRules',
  'DNASequenceBusRules',
  'DNASequencingRunBusRules',
  'ExchangeOutBusRules',
  'ExchangeOutPrepBusRules',
  'ExtractorBusRules',
  'FieldNotebookBusRules',
  'FieldNotebookPageSetBusRules',
  'FundingAgentBusRules',
  'GeographyBusRules',
  'GeologicTimePeriodBusRules',
  'GiftBusRules',
  'GiftPreparationBusRules',
  'GroupPersonBusRules',
  'InfoRequestBusRules',
  'LithoStratBusRules',
  'LoanBusRules',
  'LoanGiftShipmentBusRules',
  'LoanPreparationBusRules',
  'LoanReturnPreparationBusRules',
  'LocalityBusRules',
  'PcrPersonBusRules',
  'PermitBusRules',
  'PickListBusRules',
  'PreparationBusRules',
  'PrepTypeBusRules',
  'SpecifyUserBusRules',
  'StorageBusRules',
  'TaxonBusRules',
  'TreatmentEventBusRules',
  'TreeDefBusRules',
]);

/**
 * Most of the time business rules class name can be inferred from table name.
 * Exceptions:
 */
export const businessRulesOverride: Partial<RR<keyof Tables, string>> = {
  Shipment: 'LoanGiftShipment',
  CollectionObjectAttachment: 'Attachment',
};

export const getBusinessRuleClassFromTable = (
  tableName: keyof Tables
): string =>
  businessRulesOverride[tableName] === undefined
    ? tablesWithBusRulesIn6.has(`${tableName}BusRules`)
      ? `edu.ku.brc.specify.datamodel.busrules.${tableName}BusRules`
      : ''
    : `edu.ku.brc.specify.datamodel.busrules.${businessRulesOverride[
        tableName
      ]!}BusRules`;
