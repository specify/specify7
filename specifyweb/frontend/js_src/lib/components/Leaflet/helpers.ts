/**
 * Utility functions for getting locality data from locality resource and
 * workbench dataset
 *
 * @module
 */

import { Coord } from '../../utils/latLong';
import type { IR, RA } from '../../utils/types';
import type { MappingPath } from '../WbPlanView/Mapper';
import {
  mappingPathToString,
  valueIsTreeRank,
} from '../WbPlanView/mappingHelpers';
import { mappingLocalityColumns, requiredLocalityColumns } from './config';

export type Field<T extends Readonly<unknown>> = {
  readonly headerName: string;
  readonly value: T;
};

export type LocalityData = IR<Field<number | string>>;

export const getLocalityField = (
  row: RA<string>,
  headers: RA<string>,
  localityColumns: IR<string>,
  fieldName: string
): string => row[headers.indexOf(localityColumns[fieldName] ?? -1)] ?? '';

export function formatCoordinate(coordinate: string): number {
  if (coordinate === '' || coordinate === '0') return 0;

  const parsedCoordinate = Coord.parse(coordinate)?.toDegs();
  return parsedCoordinate === undefined
    ? 0
    : parsedCoordinate.components[0] * parsedCoordinate.sign;
}

export const findRanksInMappings = (
  mappingPaths: RA<MappingPath>
): RA<{ readonly groupName: string; readonly treeRankLocation: number }> =>
  mappingPaths
    .map((mappingPath) => ({
      mappingPath,
      treeRankLocation: mappingPath.findIndex(valueIsTreeRank),
    }))
    .map(({ mappingPath, treeRankLocation }) =>
      treeRankLocation === -1
        ? { groupName: '', treeRankLocation }
        : {
            treeRankLocation,
            groupName: mappingPathToString(
              mappingPath.slice(0, treeRankLocation)
            ),
          }
    );

export const getLocalityData = (
  localityColumns: IR<string>,
  getField: (fieldName: string) => Field<string>,
  formatCoordinate: (fieldName: string) => Field<number>
): LocalityData | false =>
  requiredLocalityColumns.every((fieldName) => getField(fieldName).value !== '')
    ? {
        ...Object.fromEntries(
          Object.keys(localityColumns)
            .filter(
              (columnName) => !mappingLocalityColumns.includes(columnName)
            )
            .map((columnName) => [columnName, getField(columnName)])
        ),
        'locality.latitude1': formatCoordinate('locality.latitude1'),
        'locality.longitude1': formatCoordinate('locality.longitude1'),
        ...(getField('locality.latitude2').value !== '' &&
        getField('locality.longitude2').value !== '' &&
        getField('locality.latlongtype').value.toLowerCase() !== 'point'
          ? {
              'locality.latitude2': formatCoordinate('locality.latitude2'),
              'locality.longitude2': formatCoordinate('locality.longitude2'),
              'locality.latlongtype': {
                value: ['line', ''].includes(
                  getField('locality.latlongtype').value.toLowerCase()
                )
                  ? 'line'
                  : 'rectangle',
                headerName: getField('locality.latlongtype').headerName,
              },
            }
          : {}),
        'locality.latlongaccuracy': isValidAccuracy(
          getField('locality.latlongaccuracy').value
        )
          ? getField('locality.latlongaccuracy')
          : { value: '', headerName: '' },
      }
    : false;

export function isValidAccuracy(
  latlongaccuracy: string | undefined
): latlongaccuracy is string {
  try {
    return (
      latlongaccuracy !== undefined &&
      !Number.isNaN(Number.parseFloat(latlongaccuracy)) &&
      Number.parseFloat(latlongaccuracy) >= 1
    );
  } catch {
    return false;
  }
}
