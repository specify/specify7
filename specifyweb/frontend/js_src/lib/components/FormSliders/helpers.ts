import { Relationship } from '../DataModel/specifyField';

export const shouldBeToOne = (
  relationship: Relationship | undefined
): boolean =>
  relationship?.type.includes('-to-many') === false ||
  (relationship?.type.includes('-to-many') === true &&
    !relationship.isDependent() &&
    relationship.getReverse()?.isDependent() === true);
