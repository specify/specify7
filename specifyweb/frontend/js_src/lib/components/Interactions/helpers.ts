import { ajax } from '../../utils/ajax';
import { formData } from '../../utils/ajax/helpers';
import type { RA, RestrictedTuple } from '../../utils/types';
import type { AnyInteractionPreparation } from '../DataModel/helperTypes';
import type { Tables } from '../DataModel/types';

export const interactionPrepTables: RestrictedTuple<
  AnyInteractionPreparation['tableName']
> = [
  'DisposalPreparation',
  'ExchangeInPrep',
  'ExchangeOutPrep',
  'GiftPreparation',
  'LoanPreparation',
];

type ExtractInteraction<T extends string> =
  T extends `${infer Prefix}Prep${string}` ? Prefix : never;

export type InteractionWithPreps = Tables[ExtractInteraction<
  AnyInteractionPreparation['tableName']
>];

export const interactionsWithPrepTables: RestrictedTuple<
  InteractionWithPreps['tableName']
> = ['Disposal', 'ExchangeIn', 'ExchangeOut', 'Gift', 'Loan'];

export type PreparationData = {
  readonly catalogNumber: string;
  readonly collectionObjectId: number;
  readonly taxon: string;
  readonly taxonId: number;
  readonly preparationId: number;
  readonly prepType: string;
  readonly countAmount: number;
  readonly loaned: number;
  readonly gifted: number;
  readonly exchanged: number;
  readonly available: number;
};

export type PreparationRow = readonly [
  catalogNumber: string,
  collectionObjectId: number,
  taxonFullName: string,
  taxonId: number,
  preparationId: number,
  prepType: string,
  preparationCountAmt: number,
  amountLoaned: string | null,
  amountedGifted: string | null,
  amountExchanged: string | null,
  amountAvailable: string
];

export const getPrepsAvailableForLoanRs = async (
  recordSetId: number,
  isLoan: boolean
) =>
  ajax<RA<PreparationRow>>(
    `/interactions/preparations_available_rs/${recordSetId}/`,
    {
      method: 'POST',
      headers: { Accept: 'application/json' },
      body: formData({ isLoan }),
    }
  ).then(({ data }) => data);

export const getPrepsAvailableForLoanCoIds = async (
  idField: string,
  collectionObjectIds: RA<string>,
  isLoan: boolean
) =>
  ajax<RA<PreparationRow>>('/interactions/preparations_available_ids/', {
    method: 'POST',
    headers: { Accept: 'application/json' },
    body: formData({
      id_fld: idField,
      co_ids: collectionObjectIds,
      isLoan,
    }),
  }).then(({ data }) => data);

export const getInteractionsForPrepId = async (prepId: number) =>
  ajax<RA<readonly [number, string | null, string | null, string | null]>>(
    '/interactions/prep_interactions/',
    {
      method: 'POST',
      headers: { Accept: 'application/json' },
      body: formData({ prepIds: prepId }),
    }
  ).then(({ data }) => data[0]);

export const getPrepAvailability = async (
  prepId: number,
  iPrepId: number | undefined,
  iPrepName: string
) =>
  ajax(
    `/interactions/prep_availability/${prepId}/${
      typeof iPrepId === 'number' ? `${iPrepId}/${iPrepName}/` : ''
    }`,
    { headers: { Accept: 'application/json' } }
  ).then(({ data }) => data);
