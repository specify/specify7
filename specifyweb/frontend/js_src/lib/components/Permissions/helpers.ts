import { getCache } from '../../utils/cache';
import { f } from '../../utils/functools';
import type { RA } from '../../utils/types';
import type { AnyTree } from '../DataModel/helperTypes';
import { schema } from '../DataModel/schema';
import type { LiteralField, Relationship } from '../DataModel/specifyField';
import type { Tables } from '../DataModel/types';
import { userInformation } from '../InitialContext/userInformation';
import { toolDefinitions } from '../Security/registry';
import { tableNameToResourceName } from '../Security/utils';
import type { tableActions } from './definitions';
import {
  getDerivedPermissions,
  getOperationPermissions,
  getTablePermissions,
} from './index';

/*
 * FEATURE: use localized action and resource names in all these log messages
 * REFACTOR: use <ProtectedTable> and etc in favor of this function
 */
/**
 * Security errors are logged so that admins can see why a particular UI
 * component is disabled or missing
 */
export function hasTablePermission(
  tableName: keyof Tables,
  action: (typeof tableActions)[number],
  collectionId = schema.domainLevelIds.collection
): boolean {
  const isReadOnly = getCache('forms', 'readOnlyMode') ?? false;
  if (isReadOnly && action !== 'read') return false;
  if (
    getTablePermissions()[collectionId][tableNameToResourceName(tableName)][
      action
    ]
  )
    return true;
  console.log(`No permission to ${action} ${tableName}`);
  return false;
}

export const hasPermission = <
  RESOURCE extends keyof ReturnType<typeof getOperationPermissions>[number],
>(
  resource: RESOURCE,
  action: keyof ReturnType<typeof getOperationPermissions>[number][RESOURCE],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  resource === '%' && action === '%'
    ? userInformation.isadmin
    : getOperationPermissions()[collectionId][resource][action]
      ? true
      : (f.log(`No permission to ${action.toString()} ${resource}`) ?? false);

export const hasToolPermission = (
  tool: keyof ReturnType<typeof toolDefinitions>,
  action: (typeof tableActions)[number],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  (toolDefinitions()[tool].tables as RA<keyof Tables>).every((tableName) =>
    hasTablePermission(tableName, action, collectionId)
  );

export const hasTreeAccess = (
  treeName: AnyTree['tableName'],
  action: (typeof tableActions)[number],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  hasTablePermission(treeName, action, collectionId) &&
  hasTablePermission(`${treeName}TreeDef`, action, collectionId) &&
  hasTablePermission(`${treeName}TreeDefItem`, action, collectionId);

export const hasDerivedPermission = <
  RESOURCE extends keyof ReturnType<typeof getDerivedPermissions>[number],
>(
  resource: RESOURCE,
  action: keyof ReturnType<typeof getDerivedPermissions>[number][RESOURCE],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  getDerivedPermissions()[collectionId][resource][action]
    ? true
    : (f.log(`No permission to ${action.toString()} ${resource}`) ?? false);

/** Check if user has a given permission for each table in a mapping path */
export const hasPathPermission = (
  mappingPath: RA<LiteralField | Relationship>,
  action: (typeof tableActions)[number],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  mappingPath
    .map((field) =>
      field.isRelationship ? field.relatedTable.name : field.table.name
    )
    .every((tableName) => hasTablePermission(tableName, action, collectionId));
