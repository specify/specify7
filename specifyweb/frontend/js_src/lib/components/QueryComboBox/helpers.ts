import type { RA, WritableArray } from '../../utils/types';
import { toTable, toTreeTable } from '../DataModel/helpers';
import type { AnySchema } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import { idFromUrl } from '../DataModel/resource';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { tables } from '../DataModel/tables';
import type { SpQuery, SpQueryField } from '../DataModel/types';
import { getMainTableFields } from '../Formatters/formatters';
import { userInformation } from '../InitialContext/userInformation';
import { userPreferences } from '../Preferences/userPreferences';
import { queryFieldFilters } from '../QueryBuilder/FieldFilter';
import { QueryFieldSpec } from '../QueryBuilder/fieldSpec';
import { flippedSortTypes } from '../QueryBuilder/helpers';
import type { TypeSearch } from './spec';
import type { CollectionRelationships } from './useCollectionRelationships';
import type { QueryComboBoxTreeData } from './useTreeData';

export function makeComboBoxQuery({
  fieldName,
  value,
  isTreeTable,
  typeSearch: { table },
  specialConditions,
}: {
  readonly fieldName: string;
  readonly value: string;
  readonly isTreeTable: boolean;
  readonly typeSearch: TypeSearch;
  readonly specialConditions: RA<SpecifyResource<SpQueryField>>;
}): SpecifyResource<SpQuery> {
  const query = new tables.SpQuery.Resource({}, { noBusinessRules: true });
  query.set('name', 'Ephemeral QueryCBX query');
  query.set('contextName', table.name);
  query.set('contextTableId', table.tableId);
  query.set('selectDistinct', false);
  query.set('countOnly', false);
  query.set('specifyUser', userInformation.resource_uri);
  query.set('isFavorite', false);
  query.set('ordinal', null);

  const searchAlgorithm = userPreferences.get(
    'form',
    'queryComboBox',
    isTreeTable ? 'treeSearchAlgorithm' : 'searchAlgorithm'
  );
  const searchField = QueryFieldSpec.fromPath(table.name, fieldName.split('.'))
    .toSpQueryField()
    .set('isDisplay', false)
    .set('startValue', searchAlgorithm === 'contains' ? `%${value}%` : value)
    .set(
      'operStart',
      searchAlgorithm === 'contains'
        ? queryFieldFilters.like.id
        : queryFieldFilters.startsWith.id
    );

  const displayField = QueryFieldSpec.fromPath(table.name, [])
    .toSpQueryField()
    .set('isDisplay', true)
    .set('sortType', flippedSortTypes.ascending);

  query.set('fields', [searchField, displayField, ...specialConditions]);

  return query;
}

export function getQueryComboBoxConditions({
  resource,
  fieldName,
  collectionRelationships,
  treeData,
  subViewRelationship,
  relatedTable,
}: {
  readonly resource: SpecifyResource<AnySchema>;
  readonly fieldName: string;
  readonly treeData: QueryComboBoxTreeData | undefined;
  readonly collectionRelationships: CollectionRelationships | undefined;
  readonly relatedTable: SpecifyTable;
  readonly subViewRelationship: Relationship | undefined;
}): RA<SpecifyResource<SpQueryField>> {
  const fields: WritableArray<SpecifyResource<SpQueryField>> = [];
  const treeResource = toTreeTable(resource);
  if (typeof treeResource === 'object') {
    const tableId = resource.specifyTable.tableId;
    // Add not-a-descendant condition
    if (!treeResource.isNew())
      fields.push(
        QueryFieldSpec.fromStringId(`${tableId}..nodeNumber`, false)
          .toSpQueryField()
          .set('isDisplay', false)
          .set('isNot', true)
          .set('operStart', queryFieldFilters.between.id)
          .set(
            'startValue',
            [
              treeResource.get('nodeNumber') ?? '',
              treeResource.get('highestChildNodeNumber') ?? '',
            ].join(',')
          )
      );
    if (fieldName === 'parent') {
      // Add rank limits
      let nextRankId = 0;
      if (typeof treeData === 'object') {
        let rankIndex =
          treeData.treeRanks.findIndex(
            // "rankId" is the original value; not updated with unsaved changes
            ({ rankId }) => rankId === treeResource.get('rankId')
          ) + 1;
        if (rankIndex !== 0) {
          while (
            rankIndex < treeData.treeRanks.length &&
            !treeData.treeRanks[rankIndex].isEnforced
          )
            rankIndex += 1;
          nextRankId = treeData.treeRanks[rankIndex - 1].rankId;
        }
      }
      const lastTreeRankId = treeData?.treeRanks.at(-1)!.rankId ?? 0;

      const lowestRankId = Math.min(
        lastTreeRankId,
        nextRankId || lastTreeRankId,
        treeData?.lowestChildRank ?? lastTreeRankId
      );
      if (lowestRankId !== 0)
        fields.push(
          QueryFieldSpec.fromStringId(`${tableId}..rankId`, false)
            .toSpQueryField()
            .set('isDisplay', false)
            .set('startValue', lowestRankId.toString())
            .set('operStart', queryFieldFilters.less.id)
        );
    } else if (fieldName === 'acceptedParent') {
      // Nothing to do
    } else if (fieldName === 'hybridParent1' || fieldName === 'hybridParent2') {
      // Nothing to do
    }
  }

  if (
    typeof collectionRelationships === 'object' &&
    fieldName === 'collectionRelType'
  )
    // Add condition for current collection
    fields.push(
      QueryFieldSpec.fromStringId(
        `${relatedTable.tableId}..collectionRelTypeId`,
        true
      )
        .toSpQueryField()
        .set('isDisplay', false)
        .set('operStart', queryFieldFilters.in.id)
        .set(
          'startValue',
          collectionRelationships[
            subViewRelationship?.name === 'leftSideRels' ? 'left' : 'right'
          ]
            .map(({ id }) => id)
            .join(',')
        )
    );
  return fields;
}

export const getRelatedCollectionId = (
  { left, right }: CollectionRelationships,
  resource: SpecifyResource<AnySchema>,
  fieldName: string
): number | undefined =>
  (fieldName === 'rightSide'
    ? left
    : fieldName === 'leftSide'
    ? right
    : undefined
  )?.find(
    ({ id }) =>
      id ===
      idFromUrl(
        toTable(resource, 'CollectionRelationship')?.get('collectionRelType') ??
          ''
      )
  )?.collection;

/**
 * If some value is currently in the input field, try to figure out which
 * field it is intended for and populate that field in the new resource.
 * Most of the time, that field is determined based on the search field
 */
export function pendingValueToResource(
  relationship: Relationship,
  typeSearch: TypeSearch | false | undefined,
  pendingValue: string
): SpecifyResource<AnySchema> {
  const mainFields = getMainTableFields(relationship.relatedTable.name);
  const typeSearchFields =
    (typeof typeSearch === 'object'
      ? typeSearch?.searchFields
          .filter(
            ([searchField]) =>
              !searchField.isRelationship &&
              searchField.table === relationship.relatedTable
          )
          .map(([field]) => field)
      : undefined) ?? [];
  const fieldName = (
    mainFields.find((field) => typeSearchFields.includes(field)) ??
    mainFields[0]
  )?.name;
  return new relationship.relatedTable.Resource(
    typeof fieldName === 'string' ? { [fieldName]: pendingValue } : {}
  );
}
