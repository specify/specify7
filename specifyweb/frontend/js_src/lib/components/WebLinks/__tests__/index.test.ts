import { requireContext } from '../../../tests/helpers';
import { webLinks } from '../index';

requireContext();

test('weblinks are fetched and parsed correctly', async () =>
  expect(
    webLinks.then((webLinks) =>
      Object.fromEntries(
        webLinks.map(
          (webLink) =>
            [
              webLink.name,
              // Get rid of symbols
              JSON.parse(JSON.stringify(webLink)),
            ] as const
        )
      )
    )
  ).resolves.toMatchSnapshot());
