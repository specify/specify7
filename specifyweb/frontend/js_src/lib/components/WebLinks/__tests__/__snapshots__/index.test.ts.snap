// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`weblinks are fetched and parsed correctly 1`] = `
{
  "AccessionNumber": {
    "description": "",
    "name": "AccessionNumber",
    "parts": [
      {
        "type": "UrlPart",
        "value": "http://specify6-test.nhm.ku.edu/echo.php?text1=",
      },
      {
        "field": [
          "[literalField Accession.accessionNumber]",
        ],
        "type": "Field",
      },
    ],
    "table": "[table Accession]",
    "usages": [],
  },
  "CTScan": {
    "description": "",
    "name": "CTScan",
    "parts": [
      {
        "type": "UrlPart",
        "value": "https://www.morphosource.org/biological_specimens/",
      },
      {
        "type": "ThisField",
      },
    ],
    "usages": [],
  },
  "FishBase": {
    "description": "Displays a FishBase page for the Genus and Species",
    "name": "FishBase",
    "parts": [
      {
        "type": "UrlPart",
        "value": "http://www.fishbase.org/Summary/",
      },
      {
        "type": "UrlPart",
        "value": "",
      },
      {
        "type": "UrlPart",
        "value": "-",
      },
      {
        "type": "UrlPart",
        "value": "",
      },
      {
        "type": "UrlPart",
        "value": ".html",
      },
    ],
    "usages": [],
  },
  "GUID": {
    "description": "",
    "name": "GUID",
    "parts": [
      {
        "type": "UrlPart",
        "value": "urn:uuid:",
      },
      {
        "type": "ThisField",
      },
    ],
    "usages": [],
  },
  "Genbank": {
    "description": "",
    "name": "Genbank",
    "parts": [
      {
        "type": "UrlPart",
        "value": "http://www.ncbi.nlm.nih.gov/nuccore/",
      },
      {
        "type": "ThisField",
      },
    ],
    "usages": [],
  },
  "MailTo": {
    "description": "Launches an email message window for sending an email.",
    "name": "MailTo",
    "parts": [
      {
        "type": "UrlPart",
        "value": "mailto:",
      },
      {
        "field": [
          "[literalField Agent.email]",
        ],
        "type": "Field",
      },
    ],
    "table": "[table Agent]",
    "usages": [
      {
        "fieldName": "email",
        "table": "[table Agent]",
      },
    ],
  },
  "Specify7": {
    "description": "",
    "name": "Specify7",
    "parts": [
      {
        "type": "UrlPart",
        "value": "https://ichthyology.specify.ku.edu/specify/bycatalog/KUI/",
      },
      {
        "type": "UrlPart",
        "value": "",
      },
      {
        "type": "UrlPart",
        "value": "/",
      },
      {
        "formatter": "Accession",
        "type": "FormattedResource",
      },
    ],
    "usages": [],
  },
  "WEBLINK": {
    "description": "",
    "name": "WEBLINK",
    "parts": [
      {
        "type": "UrlPart",
        "value": "http://specify6-test.nhm.ku.edu/echo.php?text1=",
      },
      {
        "type": "ThisField",
      },
    ],
    "usages": [],
  },
}
`;
